### PR TITLE
input: added chunk queue_limit option

### PR DESCRIPTION
**Pull request by simbou2000**
Signed-off-by: Simon Bouchard <sbouchard@rbbn.com>

Add 'queue_limit' option to limit the number of chunks an input can queue from the engine to be processed.
This is to prevent queuing all the available 'up_chunk' as 'busy_chunk' to a stalled output and unable to pack multiple following messages in newer chunks.
When no more 'up_chunk' can be created, the rapidly increasing of chunks due to creating each message in individual files cause a considerable overhead that can prevent fluent-bit from successfully recovering while in high load.

Addresses #2566
The same scenario as the opened issue greatly reduces the cpu usage due to efficient packing of messages in chunks and allows prompt processing of the current load while recovering.

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [x] Example configuration file for the change: [td-agent-bit.txt](https://github.com/fluent/fluent-bit/files/5593435/td-agent-bit.txt)
- [x] Debug log output from testing the change: [debug.zip](https://github.com/fluent/fluent-bit/files/5593557/debug.zip)
- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found: [valgrind.txt](https://github.com/fluent/fluent-bit/files/5593439/valgrind.txt)

**Documentation**

- [N/A] Documentation required for this feature

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

**Comment by simbou2000** [comment]
I checked that the errors in the continuous integration are not related to changes made by this pull request.

**Comment by edsiper** [comment]
PR review on hold until finishing the troubleshooting and fixes-poc on https://github.com/fluent/fluent-bit/issues/2566#issuecomment-735504780

File: src/flb_engine_dispatch.c [hunk 1]
```diff
@@ -239,11 +240,19 @@ int flb_engine_dispatch(uint64_t id, struct flb_input_instance *in,
     mk_list_foreach_safe(head, tmp, &in->chunks) {
         ic = mk_list_entry(head, struct flb_input_chunk, _head);
         if (ic->busy == FLB_TRUE) {
+            total_chunks_busy++;
+            continue;
+        }
+
+        /* Limit the number of chunk queued if a limit was set */
+        if (in->queue_limit > 0 && total_chunks_busy >= in->queue_limit) {
             continue;
```
Comment (manojna10) [inline src/flb_engine_dispatch.c:249]: break rather than continue?

**Comment by edsiper** [comment]
Hi, 

Looks like this kind of feature and behavior is already implemented in Fluent Bit, would you please try to reproduce the original issue you faced with Fluent Bit v1.8.8 ?

**Comment by simbou2000** [comment]
Hi @edsiper and @qingling128,
Like I said in the #2566, we were able to test with Fluent Bit 1.8.8 and appears to have the previous issues resolved.
We'll be switching to the newer official version of Fluent Bit and close this PR.
Thanks.

**Pull request closed without merging**